backend/src/bundles/chat/open-ai.service.ts

```diff
@@ -0,0 +1,84 @@
+import OpenAI, { type OpenAI as OpenAIType } from 'openai';
+import { type ValueOf } from 'shared';
+import { type Tiktoken, encoding_for_model } from 'tiktoken';
+
+import { type BaseConfig } from '~/common/config/base-config.package.js';
+
+import { CHAT_MODEL, MAX_TOKEN } from './libs/constants/constants.js';
+import { type OpenAIRole } from './libs/enums/open-ai-role.enum.js';
+import {
+    type Message,
+    type OpenAIService as OpenAIServiceModule,
+} from './libs/types/types.js';
+
+class OpenAIService implements OpenAIServiceModule {
+    private config: BaseConfig;
+    private openAi: OpenAIType;
+    private modelEncoding: Tiktoken;
+
+    public constructor(config: BaseConfig) {
+        this.config = config;
+        this.openAi = this.initOpenAi();
+        this.modelEncoding = encoding_for_model('gpt-4o-mini');
+    }
+
+    private initOpenAi = (): OpenAIType => {
+        return new OpenAI({
+            apiKey: this.config.ENV.APP.OPEN_AI_KEY,
+        });
+    };
+
+    public addMessageToHistory(
+        chatHistory: Message[],
+        userMessage: string,
+        role: ValueOf<typeof OpenAIRole>,
+    ): void {
+        const newUserMessage = {
+            content: userMessage,
+            role,
+        };
+
+        chatHistory.push(newUserMessage);
```

Comment:
Isn't it better practice to keep parameters immutable? I see that you change messages array parameter in other methods as well.

https://github.com/ryanmcdermott/clean-code-javascript?tab=readme-ov-file#avoid-side-effects-part-2